File: java/xgboost4j-demo/src/main/java/org/dmlc/xgboost4j/demo/CustomObjective.java
/*
 Copyright (c) 2014 by Contributors 

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
    
 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
package org.dmlc.xgboost4j.demo;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.dmlc.xgboost4j.Booster;
import org.dmlc.xgboost4j.IEvaluation;
import org.dmlc.xgboost4j.DMatrix;
import org.dmlc.xgboost4j.IObjective;
import org.dmlc.xgboost4j.demo.util.Params;
import org.dmlc.xgboost4j.util.Trainer;
import org.dmlc.xgboost4j.util.XGBoostError;

/**
 * an example user define objective and eval
 * NOTE: when you do customized loss function, the default prediction value is margin
 * this may make buildin evalution metric not function properly
 * for example, we are doing logistic loss, the prediction is score before logistic transformation
 * he buildin evaluation error assumes input is after logistic transformation
 * Take this in mind when you use the customization, and maybe you need write customized evaluation function
 * @author hzx
 */
public class CustomObjective {
    /**
     * loglikelihoode loss obj function
     */
    public static class LogRegObj implements IObjective {
        private static final Log logger = LogFactory.getLog(LogRegObj.class);
        
        /**
         * simple sigmoid func
         * @param input
         * @return 
         * Note: this func is not concern about numerical stability, only used as example
         */
        public float sigmoid(float input) {
            float val = (float) (1/(1+Math.exp(-input)));
            return val;
        }
        
        public float[][] transform(float[][] predicts) {
            int nrow = predicts.length;
            float[][] transPredicts = new float[nrow][1];
            
            for(int i=0; i<nrow; i++) {
                transPredicts[i][0] = sigmoid(predicts[i][0]);
            }
            
            return transPredicts;
        }

        @Override
        public List<float[]> getGradient(float[][] predicts, DMatrix dtrain) {
            int nrow = predicts.length;
            List<float[]> gradients = new ArrayList<>();
            float[] labels;
            try {
                labels = dtrain.getLabel();
            } catch (XGBoostError ex) {
                logger.error(ex);
                return null;
            }
            float[] grad = new float[nrow];
            float[] hess = new float[nrow];
            
            float[][] transPredicts = transform(predicts);
            
            for(int i=0; i<nrow; i++) {
                float predict = transPredicts[i][0];
                grad[i] = predict - labels[i];
                hess[i] = predict * (1 - predict);
            }
            
            gradients.add(grad);
            gradients.add(hess);
            return gradients;
        }        
    }
    
    /**
     * user defined eval function.
     * NOTE: when you do customized loss function, the default prediction value is margin
     * this may make buildin evalution metric not function properly
     * for example, we are doing logistic loss, the prediction is score before logistic transformation
     * the buildin evaluation error assumes input is after logistic transformation
     * Take this in mind when you use the customization, and maybe you need write customized evaluation function
     */
    public static class EvalError implements IEvaluation {
        private static final Log logger = LogFactory.getLog(EvalError.class);
        
        String evalMetric = "custom_error";
        
        public EvalError() {
        }
        
        @Override
        public String getMetric() {
            return evalMetric;
        }

        @Override
        public float eval(float[][] predicts, DMatrix dmat) {
            float error = 0f;
            float[] labels;
            try {
                labels = dmat.getLabel();
            } catch (XGBoostError ex) {
                logger.error(ex);
                return -1f;
            }
            int nrow = predicts.length;
            for(int i=0; i<nrow; i++) {
                if(labels[i]==0f && predicts[i][0]>0) {
                    error++;
                }
                else if(labels[i]==1f && predicts[i][0]<=0) {
                    error++;
                }
            }
            
            return error/labels.length;
        }
    }
    
    public static void main(String[] args) throws XGBoostError {
        //load train mat (svmlight format)
        DMatrix trainMat = new DMatrix("../../demo/data/agaricus.txt.train");
        //load valid mat (svmlight format)
        DMatrix testMat = new DMatrix("../../demo/data/agaricus.txt.test");
        
        //set params
        //set params
        Params param = new Params() {
            {
                put("eta", 1.0);
                put("max_depth", 2);
                put("silent", 1);
            }
        };
        
        //set round
        int round = 2;
        
        //specify watchList
        List<Map.Entry<String, DMatrix>> watchs =  new ArrayList<>();
        watchs.add(new AbstractMap.SimpleEntry<>("train", trainMat));
        watchs.add(new AbstractMap.SimpleEntry<>("test", testMat));
        
        //user define obj and eval
        IObjective obj = new LogRegObj();
        IEvaluation eval = new EvalError();
        
        //train a booster
        System.out.println("begin to train the booster model");        
        Booster booster = Trainer.train(param, trainMat, round, watchs, obj, eval);
    }
}

File: R-package/demo/custom_objective.R
require(xgboost)
# load in the agaricus dataset
data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')
dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label)

# note: for customized objective function, we leave objective as default
# note: what we are getting is margin value in prediction
# you must know what you are doing
watchlist <- list(eval = dtest, train = dtrain)
num_round <- 2

# user define objective function, given prediction, return gradient and second order gradient
# this is loglikelihood loss
logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}

# user defined evaluation function, return a pair metric_name, result
# NOTE: when you do customized loss function, the default prediction value is margin
# this may make buildin evalution metric not function properly
# for example, we are doing logistic loss, the prediction is score before logistic transformation
# the buildin evaluation error assumes input is after logistic transformation
# Take this in mind when you use the customization, and maybe you need write customized evaluation function
evalerror <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  err <- as.numeric(sum(labels != (preds > 0)))/length(labels)
  return(list(metric = "error", value = err))
}

param <- list(max_depth=2, eta=1, nthread = 2, silent=1, 
              objective=logregobj, eval_metric=evalerror)
print ('start training with user customized objective')
# training with customized objective, we can also do step by step training
# simply look at xgboost.py's implementation of train
bst <- xgb.train(param, dtrain, num_round, watchlist)

#
# there can be cases where you want additional information 
# being considered besides the property of DMatrix you can get by getinfo
# you can set additional information as attributes if DMatrix

# set label attribute of dtrain to be label, we use label as an example, it can be anything 
attr(dtrain, 'label') <- getinfo(dtrain, 'label')
# this is new customized objective, where you can access things you set
# same thing applies to customized evaluation function
logregobjattr <- function(preds, dtrain) {
  # now you can access the attribute in customized function
  labels <- attr(dtrain, 'label')
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}
param <- list(max_depth=2, eta=1, nthread = 2, silent=1, 
              objective=logregobjattr, eval_metric=evalerror)
print ('start training with user customized objective, with additional attributes in DMatrix')
# training with customized objective, we can also do step by step training
# simply look at xgboost.py's implementation of train
bst <- xgb.train(param, dtrain, num_round, watchlist)
